### ISSUE TITLE
Grails 4.0.3 - Micronaut MethodInterceptor does not run following re-compile

### ISSUE THREAD

**Issue by philippreston** [grails/grails-core#11531]
### Task List

- [X] Steps to reproduce provided
- [ ] Stacktrace (if present) provided
- [X] Example that reproduces the problem uploaded to Github
- [X] Full description of the issue provided (see below)

### Steps to Reproduce

1. Clone the repository at https://github.com/philippreston/micronaut-interceptor-issue
2. Start the grails application `./gradle bootRun`
3. Make a request to endpoint `http http://127.0.0.1:8080/test/request "accept:application/json"`
4. Touch a file to force re-compile `touch grails-app/services/micronaut/interceptor/issue/TestService.groovy`
5. Make a request to endpoint `http http://127.0.0.1:8080/test/request "accept:application/json"`

### Expected Behaviour

Our platform uses grails to talk to our Micronaut backend services, via the Micronaut HTTP Client.  We use a two Micronaut features in our application in order to get a token into the Micronaut HTTP Client Request:

1. MethodInterceptor - on the HttpClient to set an attribute on the MethodInvocationContext
2. HttpClientFilter - lifts the attribute from the InvocationContext on the request to get the value

This works as expected and we roughly use it as per the example application to get a token from the session into the backend request.  

Where an issue occurs is in development - where we might be running the grails app, making a change somewhere and then testing the change on the front end.  The change of file triggers a re-compile and the running app will reload.

We would expect that the MethodInterceptor would continue to be called on the execution of a HttpClient method as before

### Actual Behaviour

However... Following the reload - the MethodInterceptor no longer gets called - and as such the security details are not put onto the Request we make of the backend. 

So the MethodInterceptor looks as follows - setting a dummy value on the context

```groovy
@Singleton
class TestInterceptor implements MethodInterceptor<Object, Object> {

    final public static String AUTH_KEY = "AUTH"

    @Override
    Object intercept(MethodInvocationContext<Object, Object> context) {
        log.trace("Interceptor - Adding dummy token to context")
        context.setAttribute(AUTH_KEY, "1234567890123456789001234567890")
        return context.proceed()
    }
}
```

Our Filter looks like:

```groovy
@Filter("/**")
class TestFilter implements HttpClientFilter {

    @Override
    Publisher<? extends HttpResponse<?>> doFilter(MutableHttpRequest<?> request, ClientFilterChain chain) {


        InvocationContext context = (InvocationContext) request.getAttribute(HttpAttributes.INVOCATION_CONTEXT).orElse(null)
        if(!context) {
            log.error "Filter - No context found"
            return chain.proceed(request)
        }

        // Get the token from the context 
        String token = context.getAttribute(TestInterceptor.AUTH_KEY, String.class).orElse(null)
        if(token) {
            log.info "Filter - Found token from context -> ${token}"
            request.bearerAuth(token)
        }
        else {
            log.error "Filter - NO TOKEN ON CONTEXT - Interceptor Did Not Fire!!"
        }
        
        // Continue
        return chain.proceed(request)
    }
}
```
We can see the following logging **BEFORE** the reload

```
2020-05-05 23:23:51.703  INFO --- [nio-8080-exec-2] m.interceptor.issue.TestController       : Controller - Calling External Request
2020-05-05 23:23:51.704  INFO --- [nio-8080-exec-2] micronaut.interceptor.issue.TestService  : Service - calling backend service
2020-05-05 23:23:51.704 TRACE --- [nio-8080-exec-2] m.interceptor.issue.TestInterceptor      : Interceptor - Adding dummy token to context
2020-05-05 23:23:51.706  INFO --- [nio-8080-exec-2] micronaut.interceptor.issue.TestFilter   : Filter - Found token from context -> 1234567890123456789001234567890
2020-05-05 23:23:51.725  INFO --- [nio-8080-exec-1] m.interceptor.issue.TestController       : Dummy Microservice Controller - Getting Object (Token: Bearer 1234567890123456789001234567890)
2020-05-05 23:23:51.728  INFO --- [ntLoopGroup-1-5] micronaut.interceptor.issue.TestService  : Service - Received from backend: micronaut.interceptor.issue.TestDto(1qaz2wsx3edc, Hello World)
```
And then following the recompile

```
2020-05-05 23:24:39.633  INFO --- [nio-8080-exec-1] m.interceptor.issue.TestController       : Controller - Calling External Request
2020-05-05 23:24:39.635  INFO --- [nio-8080-exec-1] micronaut.interceptor.issue.TestService  : Service - calling backend service
2020-05-05 23:24:39.666 ERROR --- [nio-8080-exec-1] micronaut.interceptor.issue.TestFilter   : Filter - NO TOKEN ON CONTEXT - Interceptor Did Not Fire!!
2020-05-05 23:24:39.676  INFO --- [nio-8080-exec-2] m.interceptor.issue.TestController       : Dummy Microservice Controller - Getting Object (Token: null)
2020-05-05 23:24:40.106 ERROR --- [ntLoopGroup-3-1] micronaut.interceptor.issue.TestService  : Service - received HTTP/401
2020-05-05 23:24:40.275 ERROR --- [nio-8080-exec-1] o.g.web.errors.GrailsExceptionResolver   : NullPointerException occurred when processing request: [GET] /test/request
```

The MethodInterceptor will not work until restart of the grails-app.

### Environment Information

- **Operating System**:  MacOSX 
- **Grails Version:** 4.0.3
- **JDK Version:** 11.0.3
- **Container Version (If Applicable):** TODO

### Example Application

The example application is representative of an issue we are seeing in our application.  

https://github.com/philippreston/micronaut-interceptor-issue

The Micronaut HTTP Client calls the grails app itself, rather than a Micronaut service but this is enough to simulate the issue we see (using the steps above).  

Looking at the logging of BeanLifecycle - the TestInterceptor bean is created.  However if you enable AOP trace logging - you can see that it is **not** one of the interceptors that is passed into the InterceptorChain, following the reload.

Many thanks

**Comment by philippreston** [comment]
Hey 

Just wanted to check in on this and see if this was still looking good for 4.0.4 and whether there was any tentative target for that release?

Much appreciated

Phil

**Comment by puneetbehl** [comment]
I could not get a chance to look at it as a part of Grails 4.0.4.

**Comment by philippreston** [comment]
Hey Just checking in on this one.

Many thanks

Phil

**Comment by puneetbehl** [comment]
This is happening because Micornuat's [AnnotationMetadata.getAnnotationType(@Nonnull String name)](https://github.com/micronaut-projects/micronaut-core/blob/1.3.x/core/src/main/java/io/micronaut/core/annotation/AnnotationMetadata.java#L420) do not use the correct classloader. 

We need to add an additional method `getAnnotationType(name, classloader)` which also takes classloader as another argument. I will be sending a PR to Micronaut 2.1 regarding the same and this would be fixed in Grails 4.1.0.

**Comment by philippreston** [comment]
Awesome - thanks for the update.  Going to 4.1.0 is on our stack so that should all come together nicely.
